src/number.cr

```diff
@@ -87,50 +87,138 @@ struct Number
     %array
   end
 
-  # Invokes the given block with the sequence of numbers starting at `self`,
-  # incremented by *by* on each call, and with an optional *to*.
+  # Iterates from `self` to *limit* incrementing by the amount of *step* on each
+  # iteration.
   #
   # ```
-  # 3.step(to: 10, by: 2) do |n|
-  #   puts n
+  # ary = [] of Int32
+  # 1.step(to: 4, by: 2) do |x|
+  #   ary << x
   # end
+  # ary                       # => [1, 3]
+  # 1.step(to: 4, by: 2).to_a # => [1, 3]
   # ```
   #
-  # Output:
+  # The type of each iterated element is `typeof(self + step)`.
   #
-  # ```text
-  # 3
-  # 5
-  # 7
-  # 9
-  # ```
-  def step(*, to = nil, by = 1)
-    x = self + (by - by)
-
-    if to
-      if by > 0
-        while x <= to
-          yield x
-          x += by
-        end
-      elsif by < 0
-        while x >= to
-          yield x
-          x += by
-        end
+  # If *to* is `nil`, iteration is open ended.
+  #
+  # The starting point (`self`) is always iterated as first element, with two
+  # exceptions:
+  # * if `self` and *to* don't compare (i.e. `(self <=> to).nil?`). Example:
+  #   `1.0.step(Float::NAN)`
+  # * if the direction of *to* differs from the direction of `by`. Example:
+  #   `1.step(to: 2, by: -1)`
+  #
+  # In those cases the iteration is empty.
+  def step(*, to limit = nil, by step)
+    # type of current should be the result of adding `step`:
+    current = self + (step - step)
+
+    if limit == current
+      yield current
+      return
+    elsif step.zero?
+      raise ArgumentError.new("Zero step size")
+    end
+
+    direction = step.sign
+
+    if limit
+      return unless (limit <=> current).try(&.sign) == direction
+
+      yield current
+
+      while ((limit - current) <=> step) == direction
```

Comment:
Do you have benchmarks to see if these space ship comparisons didn't affect the performance?

Reply:
I didn't check any benchmark and solely focused on correctness 👍 Optimization can follow.